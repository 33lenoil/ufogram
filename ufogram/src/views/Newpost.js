import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import {useLocation} from 'react-router-dom';
import createNewPost from '../api/createNewPost';

export default function Newpost() {
    const navigate = useNavigate();
    const [file, setFile] = useState('');
    const [caption, setCaption] = useState('');
    const location = useLocation();


    const handlePost = async () => {

        // post authentication

        // 1. check if both filelink and caption is empty
        if (!file && !caption){
            return;
        }

        // 2. link authentication (if there is a link)
        if (file){
            // async function isValidImageOrVideo(url) {
            //     try {
            //       console.log("001");
            //       const response = await fetch(url, { method: 'HEAD' });
            //       console.log("002");
              
            //       if (response.ok) {
            //         const contentType = response.headers.get('Content-Type');
                    
            //         // Check if the Content-Type header indicates it's an image or video
            //         return contentType.startsWith('image/') || contentType.startsWith('video/');
            //       }
                  
            //       return false;
            //     } catch (error) {
            //       // Handle network errors or other issues
            //       return false;
            //     }
            //   }

            // async function isValidImageOrVideo(url) {
            //     const img = new Image();
            //     img.src = url;
            //     return new Promise((resolve) => {
            //       img.onload = () => resolve(true);
            //       img.onerror = () => resolve(false);
            //     });
            //   }

            async function isValidImageOrVideo(url) {
                return new Promise((resolve) => {
                  const mediaElement = document.createElement('video');
                  mediaElement.src = url;
                  
                  mediaElement.onloadeddata = () => {
                    resolve(true); // It's a valid video
                    mediaElement.remove(); // Remove the element from the DOM
                  };
                  
                  mediaElement.onerror = () => {
                    // If it fails to load as a video, check if it's an image
                    const imgElement = new Image();
                    imgElement.src = url;
              
                    imgElement.onload = () => {
                      resolve(true); // It's a valid image
                      imgElement.remove(); // Remove the element from the DOM
                    };
              
                    imgElement.onerror = () => {
                      resolve(false); // It's neither an image nor a video
                      imgElement.remove(); // Remove the element from the DOM
                    };
                  };
                });
              }
    
             // check if the image/video link to upload is valid
             try {
                const isValid = await isValidImageOrVideo(file);
                if (!isValid) {
                  console.log('Not a valid image or video URL');
                  return;
                }
              } catch (error) {
                console.error('Error:', error);
              }
        }

          
        
            //after both authentication, create a new post
            const author = {
                "id": location.state.userId,
                "username": location.state.username
            }
    
            try {
                const status = await createNewPost(caption, file, author);
                console.log('Status', status);
    
                if (status === 201) {
    
                    navigate('/main', {
                        state: {
                            userId: location.state.userId,
                            username: location.state.username,
                            users: location.state.users,
                        },
                    });
                } else {
                    return 'Error!';
                }
            } catch (error) {
                throw error;
            }

    };

    const handleMain = () => {
        navigate('/main', { state: { userId: location.state.userId, username: location.state.username, users: location.state.users } });
    };


    const handleFile = (fileEvent) => {
        setFile(fileEvent.target.value);
    }

    const handleCaption = (captionEvent) => {
        setCaption(captionEvent.target.value);
    };

    return (
        <div style={{display: "flex", justifyContent: "space-evenly"}}>
            <div style={{display: "flex", justifyContent: "center", position: "fixed", width: "100%", background: "#8769b6"}}>
                <h1>New Post</h1>
                <div>
                    <button type="button" title="Create New Post" onClick={handleMain}>Main</ button>
                </div>
            </div>
            <div style={{display: "flex", width: "100%", justifyContent: "center", marginTop: "100px", background: "#b6f486"}}>
                <label htmlFor='image/file'>Image/Video: </label>
                <input type="fileLink" name="fileLink" onChange={handleFile}/>
                <label htmlFor='caption'>Caption: </label>
                <input type="text" name="caption" onChange={handleCaption}/>
                <button type="button" title="discard" onClick={handleMain}>Discard</ button>
                <button type="button" title="post" onClick={handlePost}>Post</ button>
            </div>
        </div>
    );
}