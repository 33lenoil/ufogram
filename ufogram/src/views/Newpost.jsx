import React, { useState } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import createNewPost from '../api/createNewPost';

export default function Newpost() {
  const navigate = useNavigate();
  const [file, setFile] = useState('');
  const [caption, setCaption] = useState('');
  const location = useLocation();

  const handlePost = async () => {
    // post authentication

    // 1. check if both filelink and caption is empty
    if (!file && !caption) {
      return;
    }

    // 2. link authentication (if there is a link)
    async function isValidImageOrVideo(url) {
      return new Promise((resolve) => {
        const mediaElement = document.createElement('video');
        mediaElement.src = url;

        mediaElement.onloadeddata = () => {
          resolve(true); // It's a valid video
          mediaElement.remove(); // Remove the element from the DOM
        };

        mediaElement.onerror = () => {
          // If it fails to load as a video, check if it's an image
          const imgElement = new Image();
          imgElement.src = url;

          imgElement.onload = () => {
            resolve(true); // It's a valid image
            imgElement.remove(); // Remove the element from the DOM
          };

          imgElement.onerror = () => {
            resolve(false); // It's neither an image nor a video
            imgElement.remove(); // Remove the element from the DOM
          };
        };
      });
    }

    if (file) {
      try {
        const isValid = await isValidImageOrVideo(file);
        if (!isValid && !file.includes('youtube')) {
        //   console.log('Not a valid image or video URL');
          return;
        }
      } catch (error) {
        // console.error('Error:', error);
      }
    }

    // after both authentication, create a new post
    const author = {
      id: location.state.userId,
      username: location.state.username,
    };

    try {
      const status = await createNewPost(caption, file, author);
      //   console.log('Status', status);

      if (status === 201) {
        navigate('/main', {
          state: {
            userId: location.state.userId,
            username: location.state.username,
            users: location.state.users,
          },
        });
      } else {
        // return 'Error!';
      }
    } catch (error) {
    //   throw error;
    }
  };

  const handleMain = () => {
    navigate('/main', { state: { userId: location.state.userId, username: location.state.username, users: location.state.users } });
  };

  const handleFile = (fileEvent) => {
    setFile(fileEvent.target.value);
  };

  const handleCaption = (captionEvent) => {
    setCaption(captionEvent.target.value);
  };

  return (
    <div style={{ display: 'flex', justifyContent: 'space-evenly' }}>
      <div style={{
        display: 'flex', justifyContent: 'center', position: 'fixed', width: '100%', background: '#8769b6',
      }}
      >
        <h1>New Post</h1>
        <div>
          <button type="button" title="Create New Post" onClick={handleMain}>Main</button>
        </div>
      </div>
      <div style={{
        display: 'flex', width: '100%', justifyContent: 'center', marginTop: '100px', background: '#b6f486',
      }}
      >
        <label htmlFor="image/file">
          Image/Video:
          <input type="fileLink" name="fileLink" data-testid="linkBox" onChange={handleFile} />
        </label>
        <label htmlFor="caption">
          Caption:
          <input type="text" name="caption" data-testid="captionBox" onChange={handleCaption} />
        </label>
        <button type="button" title="discard" onClick={handleMain}>Discard</button>
        <button type="button" title="post" onClick={handlePost}>Post</button>
      </div>
    </div>
  );
}
