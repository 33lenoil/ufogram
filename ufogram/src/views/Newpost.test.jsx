/**
 * @jest-environment jsdom
 */

import React, { useState } from 'react';
import '@testing-library/jest-dom';
import {
  render, screen, fireEvent, waitFor,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  BrowserRouter as Router, Routes, Route, MemoryRouter,
} from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import Newpost from './Newpost';

const mockAxios = new MockAdapter(axios);
const users = [
  {
    username: 'zairuiy',
    password: '1234567',
    following: [
      {
        username: 'lionelhu',
        id: 2,
      },
    ],
    followers: [
      {
        username: 'lionelhu',
        id: 2,
      },
    ],
    id: 1,
  },
  {
    username: 'lionelhu',
    password: '1234567',
    following: [
      {
        username: 'zairuiy',
        id: 1,
      },
    ],
    followers: [
      {
        username: 'zairuiy',
        id: 1,
      },
    ],
    id: 2,
  },
];

mockAxios.onPost('http://localhost:3000/Posts').reply(201);

const mockedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:3001/newpost',
    state: { userId: 2, username: 'lionelhu', users },
  }),
  useNavigate: () => mockedNavigate,
}));

test('renders title', () => {
  render(
    <Router>
      {/* <Routes>
              <Route path="/signup" element={<Login />} />
          </Routes> */}
      <Newpost />
    </Router>,
  );
  const linkElement = screen.getByText(/New Post/);
  expect(linkElement).toBeInTheDocument();
});

test('renders main button', () => {
  render(
    <Router>
      {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
      <Newpost />
    </Router>,
  );
  const linkElement = screen.getByRole('button', {
    name: /Main/,
  });
  expect(linkElement).toBeInTheDocument();
});

test('renders post button', () => {
  render(
    <Router>
      {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
      <Newpost />
    </Router>,
  );
  const linkElement = screen.getByRole('button', {
    name: /Post/,
  });
  expect(linkElement).toBeInTheDocument();
});

test('renders discard button', () => {
  render(
    <Router>
      {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
      <Newpost />
    </Router>,
  );
  const linkElement = screen.getByRole('button', {
    name: /Discard/,
  });
  expect(linkElement).toBeInTheDocument();
});

test('renders image/video label', () => {
  render(
    <Router>
      {/* <Routes>
              <Route path="/signup" element={<Login />} />
          </Routes> */}
      <Newpost />
    </Router>,
  );
  const linkElement = screen.getByText(/Image\/Video:/);
  expect(linkElement).toBeInTheDocument();
});

test('renders caption label', () => {
  render(
    <Router>
      {/* <Routes>
              <Route path="/signup" element={<Login />} />
          </Routes> */}
      <Newpost />
    </Router>,
  );
  const linkElement = screen.getByText(/Caption:/);
  expect(linkElement).toBeInTheDocument();
});

/**
 * Snapshot Testing
 */

test('the component matches the snapshot', () => {
  const component = renderer.create(
    <Router>
      {/* <Routes>
        <Route path="/signup" element={<Signup />} />
    </Routes> */}
      <Newpost />
    </Router>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Tests main button', async () => {
  act(() => {
    render(
      <MemoryRouter initialEntries={['/newpost']}>
        {' '}
        {/* Set the initial route */}
        <Newpost />
      </MemoryRouter>,
    );
  });

  const loginButton = screen.getByText('Main');
  act(() => { fireEvent.click(loginButton); });

  // check if you have successfully navigated to the expected page
  // const newPageElement = screen.getByText('Signup');

  // expect(newPageElement).toBeInTheDocument();
  await waitFor(() => {
    expect(mockedNavigate).toHaveBeenCalledWith('/main', { state: { userId: 2, username: 'lionelhu', users } });
  });
  // expect(mockedNavigate).toHaveBeenCalledWith('/main');
  // await waitFor(() => expect(window.location.href).toContain('/main'));
});

test('Tests discard button', async () => {
  act(() => {
    render(
      <MemoryRouter initialEntries={['/newpost']}>
        {' '}
        {/* Set the initial route */}
        <Newpost />
      </MemoryRouter>,
    );
  });

  const loginButton = screen.getByText('Discard');
  act(() => { fireEvent.click(loginButton); });

  // check if you have successfully navigated to the expected page
  // const newPageElement = screen.getByText('Signup');

  // expect(newPageElement).toBeInTheDocument();
  await waitFor(() => {
    expect(mockedNavigate).toHaveBeenCalledWith('/main', { state: { userId: 2, username: 'lionelhu', users } });
  });
  // expect(mockedNavigate).toHaveBeenCalledWith('/main');
  // await waitFor(() => expect(window.location.href).toContain('/main'));
});

test('Tests post button', async () => {
  act(() => {
    render(
      <MemoryRouter initialEntries={['/newpost']}>
        {' '}
        {/* Set the initial route */}
        <Newpost />
      </MemoryRouter>,
    );
  });

  const fileLinkBox = screen.getByTestId('linkBox');
  // const usernameBox = screen.getByDisplayValue('usernameBox');

  act(() => {
    userEvent.type(fileLinkBox, 'https://picsum.photos/200/303');
  });

  // const passwordBox = screen.getByRole('textbox')
  const captionBox = screen.getByTestId('captionBox');

  act(() => {
    userEvent.type(captionBox, 'post10');
  });

  const postButton = screen.getByText('Post');
  act(() => { fireEvent.click(postButton); });

  // check if you have successfully navigated to the expected page
  // const newPageElement = screen.getByText('Signup');

  // expect(newPageElement).toBeInTheDocument();
  await waitFor(() => {
    expect(mockedNavigate).toHaveBeenCalledWith('/main', { state: { userId: 2, username: 'lionelhu', users } });
  });
  // expect(mockedNavigate).toHaveBeenCalledWith('/main');
  // await waitFor(() => expect(window.location.href).toContain('/main'));
});

/**
 *  Testing
 */

// test('welcome is displayed after click the registration button', async () => {
//   render(<Login />);
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByLabelText(/welcome/i);
//   expect(welcomeElement).toBeInTheDocument();
// });

// test('Username is displayed after clicking on registration button', async () => {
//   render(<Login />);
//   const usernameElement = screen.getByTestId('uname');
//   await userEvent.type(usernameElement, 'lili');
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByText(/lili/i);
//   expect(welcomeElement).toBeInTheDocument();
// });
