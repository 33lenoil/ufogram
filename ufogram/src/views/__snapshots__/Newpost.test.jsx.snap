// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component matches the snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-evenly",
    }
  }
>
  <div
    style={
      Object {
        "background": "#8769b6",
        "display": "flex",
        "justifyContent": "center",
        "position": "fixed",
        "width": "100%",
      }
    }
  >
    <h1>
      New Post
    </h1>
    <div>
      <button
        onClick={[Function]}
        title="Create New Post"
        type="button"
      >
        Main
      </button>
    </div>
  </div>
  <div
    style={
      Object {
        "background": "#b6f486",
        "display": "flex",
        "justifyContent": "center",
        "marginTop": "100px",
        "width": "100%",
      }
    }
  >
    <label
      htmlFor="image/file"
    >
      Image/Video:
      <input
        data-testid="linkBox"
        name="fileLink"
        onChange={[Function]}
        type="fileLink"
      />
    </label>
    <label
      htmlFor="caption"
    >
      Caption:
      <input
        data-testid="captionBox"
        name="caption"
        onChange={[Function]}
        type="text"
      />
    </label>
    <button
      onClick={[Function]}
      title="discard"
      type="button"
    >
      Discard
    </button>
    <button
      onClick={[Function]}
      title="post"
      type="button"
    >
      Post
    </button>
  </div>
</div>
`;
